peer/tworandomchoices/tworandomchoices.go

```diff
@@ -89,8 +89,10 @@ func (l *twoRandomChoicesList) Remove(peer peer.StatusPeer, _ peer.Identifier, p
 }
 
 func (l *twoRandomChoicesList) Choose(_ *transport.Request) peer.StatusPeer {
-	l.m.RLock()
-	defer l.m.RUnlock()
+	// Usage of a wite lock because r.random.Intn is not thread safe
+	// see: https://golang.org/pkg/math/rand/
```

Comment:
Just thought, using spinlock (busy wait with compare and swap) that generates 2 random numbers inside rlock. We know that collisions are rare and we don't need that "large" write lock really but lock namely for "rand". 

Or least use a separate Lock (Mutex), namely for "rand" (wrap it into method). Mutex will use some sort of spinlock as well but it may use some linked list to make it more fair.

Reply:
Thanks for the suggestion - I will investigate for this and create some benchmark before committing to it.